### ISSUE TITLE
Inline blot similar to Link

### ISSUE THREAD

**Issue by danielschwartz** [quilljs/quill#1232]
Hi All,

I'm trying to create a custom blot that is similar to Link in that typing after it doesn't keep the formatting (e.g. typing after the blot doesn't keep the same format, but reverts to no formatting), but typing inside of does. My code is below, which follow the Link blot code almost exactly (except for adding properties other than href or blocking certain formatting options), any idea why I can't get this to work?

```js
import Inline from '../blots/inline';


class UserMention extends Inline {
    static create(value) {
        let node = super.create(value);
        node.setAttribute('data-user-id', value.userId);
        node.setAttribute('data-user-name', value.userName);
        return node;
    }

    static formats(domNode) {
        return {
            userName: domNode.getAttribute('data-user-name'),
            userId: domNode.getAttribute('data-user-id')
        }
    }
}

UserMention.blotName = 'user-mention';
UserMention.className = 'user-mention'
UserMention.tagName = 'SPAN';

export default UserMention;
```

**Comment by think2011** [comment]
I also want to know.

**Comment by jhchen** [comment]
The behavior is natively specific to the `<a>` tag (see https://jsfiddle.net/x7g357ze/) and differs between platforms.

**Comment by think2011** [comment]
@jhchen So.. how can I created Inline instance similar Link or 
![11](https://cloud.githubusercontent.com/assets/3961388/21839205/751bab42-d811-11e6-8dba-c33087383bbb.gif)

I really want to know, please help me
http://codepen.io/think2011/pen/ggPKXE

**Comment by danielschwartz** [comment]
@think2011 if it's a feature that native specific to the A tag, then you will have to use a modified A tag for whatever custom blot you'd like. I'm sure there's a way to hack into Quill core, but that feels like a heavy modification just to get this behavior.

**Comment by think2011** [comment]
@danielschwartz  if I click cancel format and start input, is the result I want.
![11](https://cloud.githubusercontent.com/assets/3961388/21839205/751bab42-d811-11e6-8dba-c33087383bbb.gif)

 so I try to read source code [formats/bold.js](https://github.com/quilljs/quill/blob/develop/formats/bold.js), but I did not found any effected code..😳

**Comment by danielschwartz** [comment]
If you look at the standard inline blot format from Parchment (which Bold is a subclass of), there is nothing that dictates when it is on or off. Quill core (the part of quill that binds to the contenteditable div) is the part that affects whether a format is currently enabled or not. 

You could technically hook into the SELECTION_CHANGE event and look at the node immediately preceding the cursor, and if it matches your custom blot, disable the format by using the Quill API, but again, that feels like a very heavy way to accomplish this when you could simply base your Blot of the A tag and have it happen automatically. 

--

Daniel Schwartz

On Wed, Jan 11, 2017 at 9:29 AM "曾浩"

<
mailto:
> wrote:

a, pre, code, a:link, body { word-wrap: break-word !important; }

https://github.com/danielschwartz
if I click cancel format and start input, is the result I want.
https://cloud.githubusercontent.com/assets/3961388/21839205/751bab42-d811-11e6-8dba-c33087383bbb.gif

so I try to read source code
https://github.com/quilljs/quill/blob/develop/formats/bold.js
, but I did not found any effected code..

😳

—

You are receiving this because you were mentioned.

Reply to this email directly,
https://github.com/quilljs/quill/issues/1232#issuecomment-271882407
, or

**Comment by think2011** [comment]
Now I used `SELECTION_CHANGE` and `TEXT_CHANGE` to remove the format by custom blot, really a very heavy way to accomplish 😅

**Comment by think2011** [comment]
```html
<span style="display: inline-block;"><div>test</div></span>
```

@danielschwartz I found the simple way

**Comment by puopg** [comment]
I think I might be able to help you with this. I too am working on a facebook like mention autocomplete. What I am doing is essentially controlling the inputs explicitly. Basically, you need to hook into key events before quill gets them and also control inserting of keys yourself. 

So lets say you want to insert a character after you have added a mention with some styles of course. Use the keypress event to determine if an actual character is about to go into the editor and then prevent the default action from happening and insert that item yourself. You know that you want to insert it with no styles, you don't want quill's default action on carrying over styles.  

Now for a few other events like enter, delete, backspace. Those you need to attach to the config so that your handler runs before quill can. This way if there's something like you want to delete a character from a mention and then you want to remove all formatting from the entire mention, you can do that when you intercept the event first, then allow quill to do its normal action of deleting the character or whatnot.

**Comment by scottfr** [comment]
I'm trying to solve this too.

@danielschwartz, did using the 'a' tag work for you? In my test using tagName = 'a' doesn't result in the correct behavior and the styling is still extended.

@think2011, do you have an example of the selection_change/text_change approach?

**Comment by joeribakker** [comment]
It seems to me that it has something to do with either the cursor's behaviour:

https://github.com/quilljs/quill/blob/master/blots/cursor.js#L66

Or (or rather, maybe in combination with) Inline's `order` attribute:

https://github.com/quilljs/quill/blob/c9a2e9f4d0bbeeae20f15dd0bdad143a241a5275/blots/inline.js#L43

Changing a linkblot's name to `code` yields the same behaviour, while changing it to something arbitrary (even though I'm keeping the `A` tag name) makes the format expand rather than stop with inserting new characters after its end.

I'm not quite sure how to manipulate these functionalities in order to add my own format that will behave like a link; I've added my own format's `blotName` at the end of the `order` list, but that doesn't cut it.

**Comment by joeribakker** [comment]
I've been plugging away at the same problem, and based on your answers I implemented the following to make something behave more like a link:

```
quill.on('text-change', (delta, oldDelta, source) => {
	if (source === 'user') {
		if (
			quill.getFormat(quill.getSelection().index).like-link
			&& !quill.getFormat(quill.getSelection().index + 1).like-link
			&& delta.ops[delta.ops.length - 1].insert
		) {
			quill.removeFormat(quill.getSelection().index - 1, quill.getSelection().index);
		}
	}
});
```

It specifically looks for text changes done by the `'user'`, and only removes the format of the newly inserted character when the `delta` indicates an `insert`.

**Comment by dkreft** [comment]
I'd like to add a little extra flavor for those who are wondering what's going on under the covers here...

I tried to create a subclass of `Inline` with `tagName` of "A" and it didn't work. After much digging around, I added a `console.log('mutations: %o', mutations)` here: https://github.com/quilljs/parchment/blob/master/src/blot/scroll.ts#L27-L29 and discovered that when typing after an inserted link (in Chrome), `mutations` contains:

![screen shot 2018-09-14 at 3 35 41 pm](https://user-images.githubusercontent.com/272745/45578099-56456a00-b834-11e8-9507-98926d9bcb02.png)

When my class has a `tagName` of "A" and no other attributes, the `mutations` contains this:

![screen shot 2018-09-14 at 3 36 23 pm](https://user-images.githubusercontent.com/272745/45578122-7aa14680-b834-11e8-999a-82e5fdb6d458.png)

It's only when I added an `href` attribute to the `<a>` that it finally gave me the desired behavior.

HOWEVER, in Firefox, there's only one `MutationRecord` in `mutations` even when typing after an inserted link, so not even using an `href` attribute is going to save you in that browser.

**TL;DR**: Works in Chrome if you make sure your `<a>` has an `href` attribute; does not work in Firefox either way:

```js
export default MyCustomLinkTag extends Inline {
  static tagName = 'A'

    static create(tagName) {
      const node = super.create()

      node.setAttribute('href', 'http://xxxxx')

      return node
  }
}
```

**Comment by dkreft** [comment]
As a followup to my previous post, it would see that this behavior we're seeing here is a combination of browser-specific behavior and possibly a deficiency/bug in Quill. If you look at the basic example provided for ProseMirror (https://prosemirror.net/examples/basic/), it behaves as expected (i.e. typing immediately after a link does not add text between the `<a></a>` tags) in *both* Chrome and Firefox. I haven't dug into how ProseMirror makes this possible.

**Comment by Haemp** [comment]
We really do need a native quill way to handle the cursor behavior at the end of a starting inline blot. Treating the inline blot as a link **does** work, but has the limitations of no long being able to nest those blots since browsers won't allow nested anchor tags.

**Comment by mitya777** [comment]
You can wrap anchor tags in an object tag with a fake mime type to nest links, here are some resources to that effect https://knovigator.com/quest/nested-links-html-anchor-tags-ygqql91g

**Comment by Krypternite** [comment]
> Now I used `SELECTION_CHANGE` and `TEXT_CHANGE` to remove the format by custom blot, really a very heavy way to accomplish 😅

Would like to know how to do this
